pkgs/applications/graphics/fiji/default.nix

```diff
@@ -0,0 +1,72 @@
+{ stdenv
+, lib
+, fetchurl
+, makeWrapper
+, autoPatchelfHook
+, jdk11
+, makeDesktopItem
+, copyDesktopItems
+, runtimeShell
+}:
+stdenv.mkDerivation rec {
+  pname = "fiji";
+  version = "20201104-1356";
+
+  src = fetchurl {
+    url = "https://downloads.imagej.net/${pname}/archive/${version}/${pname}-nojre.tar.gz";
+    sha256 = "1jv4wjjkpid5spr2nk5xlvq3hg687qx1n5zh8zlw48y1y09c4q7a";
+  };
+
+  dontBuild = true;
+
+  nativeBuildInputs = [ autoPatchelfHook makeWrapper copyDesktopItems ];
+  buildInputs = [ stdenv.cc.cc.lib ];
+
+  desktopItems = [
+    (makeDesktopItem {
+      name = "fiji";
+      exec = "fiji %F";
+      icon = "fiji";
+      mimeType = "image/*;";
+      comment = "Scientific Image Analysis";
+      desktopName = "Fiji Is Just ImageJ";
+      genericName = "Fiji Is Just ImageJ";
+      categories = "Education;Science;ImageProcessing;";
+      terminal = false;
+      startupNotify = true;
+      extraEntries = ''
+        Version=1.0
+        TryExec=fiji
+        X-GNOME-FullName=Fiji Is Just ImageJ
+        StartupWMClass=fiji-Main
+      '';
+    })
+  ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/{bin,fiji,share/pixmaps}
+
+    cp -R * $out/fiji
+    rm -f $out/fiji/jars/imagej-updater-*.jar
+
+    # Disgusting hack to stop a local desktop entry being created
+    makeWrapper ${runtimeShell} $out/bin/fiji \
```

Comment:
I don't understand why this hack is needed. Can't you simply delete the `.desktop` file instead from `$out/share` :thinking: ?

Reply:
If only... `ImageJ-linux64` tries to be "helpful" by creating one in `~/.local/share`, and there's no way to turn it off short of patching it out.

This wouldn't be *that* bad, but it uses the full `/nix/store/....` path in it. Now not only does the user have 2 Fiji desktop entries, one may be broken by an update.

Reply:
Just to make it clear, by running it in a shell it doesn't create this file?

Reply:
The `--dry-run` causes it to print the command that it would run, without actually doing anything. I invoke it with `--dry-run` to get this command, then `exec` it.

```
$ ./result/fiji/ImageJ-linux64 --dry-run
java -Dpython.cachedir.skip=true -Dplugins.dir=/nix/store/xy6fykx0ddr9kzxgzib9j1y3zc4yzv2c-fiji-20201104-1356/fiji -Xmx5381m -Dimagej.splash=true -Djava.class.path=/nix/store/xy6fykx0ddr9kzxgzib9j1y3zc4yzv2c-fiji-20201104-1356/fiji/jars/imagej-launcher-6.0.0.jar -Dimagej.dir=/nix/store/xy6fykx0ddr9kzxgzib9j1y3zc4yzv2c-fiji-20201104-1356/fiji -Dij.dir=/nix/store/xy6fykx0ddr9kzxgzib9j1y3zc4yzv2c-fiji-20201104-1356/fiji -Dfiji.dir=/nix/store/xy6fykx0ddr9kzxgzib9j1y3zc4yzv2c-fiji-20201104-1356/fiji -Dfiji.executable=./result/fiji/ImageJ-linux64 -Dij.executable=./result/fiji/ImageJ-linux64 -Djava.library.path=/nix/store/6sfs2ns86fijzkm1phqrs49qx5bz0z0s-sane-config/lib/sane:/nix/store/rcnb5lclnj3vs6q8na47f038vzhaa3bv-pipewire-0.3.36-jack/lib:/nix/store/xy6fykx0ddr9kzxgzib9j1y3zc4yzv2c-fiji-20201104-1356/fiji/lib/linux64:/nix/store/xy6fykx0ddr9kzxgzib9j1y3zc4yzv2c-fiji-20201104-1356/fiji/mm/linux64 -Dscijava.context.strict=false -Dpython.console.encoding=UTF-8 net.imagej.launcher.ClassLauncher -ijjarpath jars -ijjarpath plugins net.imagej.Main
```

I also just realise the hack as-is doesn't pass through user parameters at all. I need to change it to something like this:
```nix

     # Disgusting hack to stop a local desktop entry being created
    cat <<EOF > $out/bin/.fiji-launcher-hack
    #!${runtimeShell}
    exec \$($out/fiji/ImageJ-linux64 --dry-run "\$@")
    EOF
    chmod +x $out/bin/.fiji-launcher-hack

    makeWrapper $out/bin/.fiji-launcher-hack $out/bin/fiji \
      --prefix PATH : ${lib.makeBinPath [ jdk11 ]} \
      --set JAVA_HOME ${jdk11.home}
```

Reply:
I also can't bypass the launcher as it has parameters that users have come to expect, e.g. `--headless`